juju/juju
# [JUJU-290] Engine metrics

The following exposes the number of starts a worker has for a given
model. To reduce cardinality the metrics are split into the following:

 1. Number of starts per worker globally.
 2. Number of starts of workers by model.

We do it this way to prevent the explosion of larger controllers with
lots of models dealing with lots of metrics. Hopefully these metrics
will then help operators of Juju see if there are problems with
operating Juju from things like grafana.

## QA steps

```sh
$ juju bootstrap lxd test1
$ juju add-user prometheus
$ juju change-user-password prometheus
$ juju grant prometheus read controller
```

Read the metrics endpoint to ensure that we've got the right number of workers per model.

```sh
$ juju show-controller test1 --format=json | jq '.test1.details."api-endpoints"[0]' | xargs -I % echo "https://%/introspection/metrics" | xargs curl -k -s -u user-prometheus:$PASSWORD | grep dependency
# HELP juju_dependency_engine_worker_model_starts Current number of worker starts in the dependency engine by model
# TYPE juju_dependency_engine_worker_model_starts gauge
juju_dependency_engine_worker_model_starts{model_uuid="609bda37-bf92-4b58-881a-0ea2675859d3"} 106
juju_dependency_engine_worker_model_starts{model_uuid="a2b4d60e-d23f-4af3-8403-582308fc48e8"} 33
juju_dependency_engine_worker_model_starts{model_uuid="dc2a7de6-0d7d-44e4-8c69-f6042f50b1c4"} 33
# HELP juju_dependency_engine_worker_start Current number of worker starts in the dependency engine by worker
# TYPE juju_dependency_engine_worker_start gauge
juju_dependency_engine_worker_start{worker="action-pruner"} 3
juju_dependency_engine_worker_start{worker="agent"} 4
juju_dependency_engine_worker_start{worker="agent-config-updater"} 1
juju_dependency_engine_worker_start{worker="api-address-updater"} 1
juju_dependency_engine_worker_start{worker="api-caller"} 4
juju_dependency_engine_worker_start{worker="api-config-watcher"} 4
juju_dependency_engine_worker_start{worker="api-server"} 1
juju_dependency_engine_worker_start{worker="application-scaler"} 3
juju_dependency_engine_worker_start{worker="audit-config-updater"} 1
juju_dependency_engine_worker_start{worker="broker-tracker"} 1
juju_dependency_engine_worker_start{worker="central-hub"} 1
juju_dependency_engine_worker_start{worker="certificate-updater"} 1
juju_dependency_engine_worker_start{worker="certificate-watcher"} 1
juju_dependency_engine_worker_start{worker="charm-revision-updater"} 3
juju_dependency_engine_worker_start{worker="clock"} 4
juju_dependency_engine_worker_start{worker="compute-provisioner"} 3
juju_dependency_engine_worker_start{worker="controller-port"} 1
juju_dependency_engine_worker_start{worker="deployer"} 1
juju_dependency_engine_worker_start{worker="disk-manager"} 1
juju_dependency_engine_worker_start{worker="environ-tracker"} 3
juju_dependency_engine_worker_start{worker="external-controller-updater"} 1
juju_dependency_engine_worker_start{worker="fan-configurer"} 1
juju_dependency_engine_worker_start{worker="firewaller"} 3
juju_dependency_engine_worker_start{worker="host-key-reporter"} 1
juju_dependency_engine_worker_start{worker="http-server"} 1
juju_dependency_engine_worker_start{worker="http-server-args"} 1
juju_dependency_engine_worker_start{worker="instance-mutater"} 4
juju_dependency_engine_worker_start{worker="instance-poller"} 3
juju_dependency_engine_worker_start{worker="is-controller-flag"} 1
juju_dependency_engine_worker_start{worker="is-primary-controller-flag"} 1
juju_dependency_engine_worker_start{worker="is-responsible-flag"} 3
juju_dependency_engine_worker_start{worker="lease-clock-updater"} 1
juju_dependency_engine_worker_start{worker="lease-manager"} 1
juju_dependency_engine_worker_start{worker="log-forwarder"} 3
juju_dependency_engine_worker_start{worker="log-sender"} 1
juju_dependency_engine_worker_start{worker="logging-config-updater"} 4
juju_dependency_engine_worker_start{worker="machine-action-runner"} 1
juju_dependency_engine_worker_start{worker="machine-undertaker"} 3
juju_dependency_engine_worker_start{worker="machiner"} 1
juju_dependency_engine_worker_start{worker="metric-worker"} 3
juju_dependency_engine_worker_start{worker="mgo-txn-resumer"} 1
juju_dependency_engine_worker_start{worker="migration-fortress"} 4
juju_dependency_engine_worker_start{worker="migration-inactive-flag"} 4
juju_dependency_engine_worker_start{worker="migration-master"} 3
juju_dependency_engine_worker_start{worker="migration-minion"} 1
juju_dependency_engine_worker_start{worker="model-cache"} 1
juju_dependency_engine_worker_start{worker="model-cache-initialized-flag"} 2
juju_dependency_engine_worker_start{worker="model-cache-initialized-gate"} 1
juju_dependency_engine_worker_start{worker="model-upgrade-gate"} 3
juju_dependency_engine_worker_start{worker="model-upgraded-flag"} 6
juju_dependency_engine_worker_start{worker="model-upgrader"} 6
juju_dependency_engine_worker_start{worker="model-worker-manager"} 1
juju_dependency_engine_worker_start{worker="multiwatcher"} 1
juju_dependency_engine_worker_start{worker="not-alive-flag"} 3
juju_dependency_engine_worker_start{worker="not-dead-flag"} 3
juju_dependency_engine_worker_start{worker="peer-grouper"} 1
juju_dependency_engine_worker_start{worker="presence"} 1
juju_dependency_engine_worker_start{worker="proxy-config-updater"} 1
juju_dependency_engine_worker_start{worker="pubsub-forwarder"} 1
juju_dependency_engine_worker_start{worker="raft"} 1
juju_dependency_engine_worker_start{worker="raft-backstop"} 1
juju_dependency_engine_worker_start{worker="raft-clusterer"} 1
juju_dependency_engine_worker_start{worker="raft-forwarder"} 1
juju_dependency_engine_worker_start{worker="raft-leader-flag"} 2
juju_dependency_engine_worker_start{worker="raft-transport"} 1
juju_dependency_engine_worker_start{worker="reboot-executor"} 1
juju_dependency_engine_worker_start{worker="remote-relations"} 3
juju_dependency_engine_worker_start{worker="restore-watcher"} 1
juju_dependency_engine_worker_start{worker="ssh-authkeys-updater"} 1
juju_dependency_engine_worker_start{worker="ssh-identity-writer"} 1
juju_dependency_engine_worker_start{worker="state"} 1
juju_dependency_engine_worker_start{worker="state-cleaner"} 3
juju_dependency_engine_worker_start{worker="state-config-watcher"} 1
juju_dependency_engine_worker_start{worker="status-history-pruner"} 3
juju_dependency_engine_worker_start{worker="storage-provisioner"} 4
juju_dependency_engine_worker_start{worker="termination-signal-handler"} 1
juju_dependency_engine_worker_start{worker="tools-version-checker"} 1
juju_dependency_engine_worker_start{worker="transaction-pruner"} 1
juju_dependency_engine_worker_start{worker="unconverted-api-workers"} 1
juju_dependency_engine_worker_start{worker="unit-assigner"} 3
juju_dependency_engine_worker_start{worker="upgrade-check-flag"} 2
juju_dependency_engine_worker_start{worker="upgrade-check-gate"} 1
juju_dependency_engine_worker_start{worker="upgrade-database-flag"} 1
juju_dependency_engine_worker_start{worker="upgrade-database-gate"} 1
juju_dependency_engine_worker_start{worker="upgrade-database-runner"} 1
juju_dependency_engine_worker_start{worker="upgrade-series"} 1
juju_dependency_engine_worker_start{worker="upgrade-steps-flag"} 1
juju_dependency_engine_worker_start{worker="upgrade-steps-gate"} 1
juju_dependency_engine_worker_start{worker="upgrade-steps-runner"} 1
juju_dependency_engine_worker_start{worker="upgrader"} 1
juju_dependency_engine_worker_start{worker="valid-credential-flag"} 4
```
